Update documentation around VEP and snpEff cache (#1222)
nf-core/sarek: docs/usage.md

| pon_tbi               | Mutect2                                                                                                                                                                                                                                                                                                                                                                                                                                              | [GATKBundle](https://console.cloud.google.com/storage/browser/_details/genomics-public-data/resources/broad/hg38/v0/) | https://gatk.broadinstitute.org/hc/en-us/articles/360035890631-Panel-of-Normals-PON- |

## How to customise snpEff and VEP annotation

SNPeff and VEP require a large resource of files known as a cache. These are folders composed of multiple gigabytes of files which need to be available for the software to properly function. To use these, supply the parameters `--vep_cache` and/or `--snpeff_cache` to the relevant folders.

### Specify the cache location

> Params `--snpeff_cache` and `--vep_cache` are to used to specify the locations to the root of the annotation cache folder. The cache will be located within a subfolder with the path `${vep_species}/${vep_genome}_${vep_cache_version}` for VEP and `${snpeff_species}.${snpeff_version}` for snpEff. If this directory is missing, Sarek will raise an error.

Reviewer: ```suggestion
Params `--snpeff_cache` and `--vep_cache` are to used to specify the locations to the root of the annotation cache folder. In the cache will be located a subfolder with the path `${vep_species}/${vep_genome}_${vep_cache_version}` for VEP and `${snpeff_species}.${snpeff_version}` for snpEff. If this directory is missing, Sarek will raise an error.
```

> Params `--snpeff_cache` and `--vep_cache` are to used to specify the locations to the root of the annotation cache folder. The cache will be located within a subfolder with the path `${vep_species}/${vep_genome}_${vep_cache_version}` for VEP and `${snpeff_species}.${snpeff_version}` for snpEff. If this directory is missing, Sarek will raise an error.

Reviewer 2: ```suggestion
Params `--snpeff_cache` and `--vep_cache` are used to specify the locations to the root of the annotation cache folder. The cache will be located within a subfolder with the path `${vep_species}/${vep_genome}_${vep_cache_version}` for VEP and `${snpeff_species}.${snpeff_version}` for snpEff. If this directory is missing, Sarek will raise an error.
```

> For example this is a typical folder structure for GRCh38 and WBCel235, with SNPeff version 105 and VEP version 106:

Reviewer 2: where does the 106 come from?

Reviewer: Good point, these are not the tools' version, but the tools' cache version

snpeff_genome     = 'GRCh38'
vep_genome        = 'GRCh38'
vep_species       = 'homo_sapiens'
vep_cache_version = '110'
```

> ### Usage recommendation with AWS iGenomes

Reviewer: ```suggestion
### Usage recommendation with AWS Annotation Cache
```

> Annotation cache is a resource separated from AWS iGenomes, which as its own structure and a frequent update cycle, so it is not recommended to put any annotation cache in your local AWS iGenomes folder.

Reviewer 2: ```suggestion
Annotation cache is a resource separated from AWS iGenomes, which has its own structure and a frequent update cycle, so it is not recommended to put any annotation cache in your local AWS iGenomes folder.
```

A classical organisation on a shared storage area might be:

```bash
/data/igenomes/
/data/cache/ensemblvep
/data/cache/snpeff
```

which can then be used this way in Sarek:

```bash
nextflow run nf-core/sarek \
    --igenomes_base /data/igenomes/ \
    --snpeff_cache /data/cache/snpeff/ \
    --vep_cache /data/cache/ensemblvep/ \
    ...
```

Alternatively the data may be stored on AWS S3 storage, therefore the parameters might be:

```bash
s3://my-reference-data/igenomes/
s3://my-reference-data/cache/ensemblvep/
s3://my-reference-data/cache/snpeff/
```

which can then be used this way in sarek:

```bash
nextflow run nf-core/sarek \
    --igenomes_base s3://my-reference-data/igenomes/ \
    --snpeff_cache s3://my-reference-data/cache/ensemblvep/ \
    --vep_cache s3://my-reference-data/cache/snpeff/ \
    ...
```

These params can be specified in a config file or in a profile using the params scope, or even in a json or a yaml file using the `-params-file` nextflow option.

### Use pre-populated cache for SNPeff and VEP

> Sarek stores a mirror of some cache files on AWS S3 which are set as the default parameters. They contain some genome builds which can be found by checking the contents of the S3 bucket.

Reviewer 2: maybe this section can come right after ##Usage recommendation with AWS iGenomes

```bash
vep_cache = s3://annotation-cache/vep_cache/
```

The contents of the VEP cache can be listed with the following command using the S3 CLI:

```bash
aws s3 --no-sign-request ls s3://annotation-cache/vep_cache/
```

> Simialrly, the SNPeff cache is stored at the following location on S3:

Reviewer 2: ```suggestion
Similarly, the SNPeff cache is stored at the following location on S3:
```

```bash
snpeff_cache = s3://annotation-cache/snpeff_cache/
```

The contents can be listed as follows:

```bash
aws s3 --no-sign-request ls s3://annotation-cache/snpeff_cache
```

> Be sure the specified species, genome and build matches the directory structure within the cache!

Reviewer 2: Don't 100% get it here: basically if you download you want to emulate the filder structure you describe above right?

### Use Sarek to download cache and annotate in one go

> Use the params `--download_cache`, and specify with `--tools` for which annotation tool you need to download the cache (`snpeff` and or `vep`)

Reviewer 2: download cache always downloads from annotation cache or vep/snpeff?

Reviewer: There is no cache in igenomes, but in annotation-cache.
What we have in igenomes.config are the value for the build/species/genome.
Using this possibility instead of using annotation-cache allows for people to download other species/version, since we only have c Elegans and H sapiens in annotation-cache for now

Sarek will automatically download the cache, use the biocontainers container for said tools, and use it to annotate any vcfs produced.

### Only download cache

Using the params `--build_only_index` allow for only downloading the cache for the specified tools.

### Location for the cache

Cache can be downloaded in the specified `--outdir_cache` location.
Else, it will be downloaded in `cache/` in the specified `--outdir` location.

To download cache on a cloud infrastructure, an absolute path is needed.

### Using the nf-core containers with pre-downloaded cache

> For common genomes, the VEP and snpEff parameters are pre-populated in the [igenomes.config](https://github.com/nf-core/sarek/blob/master/conf/igenomes.config) file and as such the cache files are included with the specified iGenome. However, these are infrequently updated and therefore using a pre-populated cache is preferable.

Reviewer 2: ```suggestion
For common genomes, the VEP and snpEff parameters are pre-populated in the [igenomes.config](https://github.com/nf-core/sarek/blob/master/conf/igenomes.config) file and as such the cache files are included with the specified iGenome. However, these are infrequently updated and therefore using the newest available cache is preferable.
```
or sth like that?
I understood it as not using the pre-populated cache is preferred to get the newest of the newest?

An associated Docker container for some caches is available which includes the cache and can be used. However, these are very large and can cause frequent problems so should be avoided and are preserved here for legacy reasons. The containers are only created for some species and some cache/tools versions combinations and match the tags of the Docker containers (cf DockerHub tags for these containers [`nfcore/snpeff`](https://hub.docker.com/r/nfcore/snpeff/tags) and [`nfcore/vep`](https://hub.docker.com/r/nfcore/vep/tags).

> ### Create containers with pre-downloaded cache

Reviewer 2: Maybe a link to the docs on how to specify your own container? Also when or when not to build your own container? From what I understood now you don't necessary need to build one that contains the cache?

Reviewer: added some warning and disclaimers

These params can be specified in a config file or in a profile using the params scope, or even in a json or a yaml file using the `-params-file` nextflow option.

> Note: we only recommend to store annotation data on S3 if you only store a specific version of the cache for a species, and not the whole cache.

Reviewer: sarek handles it

> Note: we only recommend to store annotation data on S3 if you only store a specific version of the cache for a species, and not the whole cache.

Reviewer: yeah no, I know why I didn't push for it, it's a mess to do it with the current modules.

> Note: we only recommend to store annotation data on S3 if you only store a specific version of the cache for a species, and not the whole cache.

Reviewer: yeah so handling this is a mess since both vep and snpeff need the path to the root folder.
Which is why I added this extra layer of folder in annotation-cache.
